controllers/component_pipelinerun_nudge_controller.go

```diff
@@ -0,0 +1,220 @@
+	"time"
+
+	"github.com/go-logr/logr"
+	applicationapiv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	contextTimeout = 300 * time.Second
+	// PipelineRunTypeLabel contains the type of the PipelineRunTypeLabel.
+	PipelineRunTypeLabel = "pipelines.appstudio.openshift.io/type"
+	// PipelineRunBuildType is the type denoting a build PipelineRun.
+	PipelineRunBuildType         = "build"
+	NudgeProcessedAnnotationName = "build.appstudio.openshift.io/component-nudge-processed"
+
+	// PipelineRunComponentLabel is the label denoting the application.
+	PipelineRunComponentLabel = "appstudio.openshift.io/component"
+	NudgeFinalizer            = "appstudio.openshift.io/build-nudge-finalizer"
+
+	PacEventType     = "pipelinesascode.tekton.dev/event-type"
+	PacEventPush     = "push"
+	ComponentNudged  = "ComponentNudged"
+	ImageUrlParam    = "IMAGE_URL"
+	ImageDigestParam = "IMAGE_DIGEST"
+)
+
+// ComponentNudgeReconciler reconciles a PipelineRun object
+type ComponentNudgeReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	EventRecorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=appstudio.redhat.com,resources=components,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=appstudio.redhat.com,resources=components/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=tekton.dev,resources=pipelineruns,verbs=get;list;watch;create;update;patch;delete;deletecollection
+// +kubebuilder:rbac:groups=tekton.dev,resources=pipelineruns/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=tekton.dev,resources=pipelineruns/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+func (r *ComponentNudgeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithTimeout(ctx, contextTimeout)
+	defer cancel()
+	log := ctrllog.FromContext(ctx).WithName("ComponentNudge").WithValues("namespace", req.NamespacedName.Namespace, "resource", req.Name)
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	pipelineRun := &tektonapi.PipelineRun{}
+	err := r.Get(ctx, req.NamespacedName, pipelineRun)
+	if err != nil {
+		log.Error(err, "Failed to get pipelineRun")
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	component, err := GetComponentFromPipelineRun(r.Client, ctx, pipelineRun)
+	if err != nil {
+		log.Error(err, "failed to get component")
+		return ctrl.Result{}, err
+	}
+
+	if len(component.Spec.BuildNudgesRef) == 0 {
+		log.Info(fmt.Sprintf("component %s has no BuildNudgesRef set", component.Name))
+		return ctrl.Result{}, err
+	}
+
+	if pipelineRun.IsDone() || pipelineRun.DeletionTimestamp != nil {
```

Comment:
Pipeline can be deleted before completion.
Wouldn't it be better to analyze `DeletionTimestamp` at the beginning of the reconcile?

Reply:
We need to remove the finalizer no matter what if the PR is deleted, deleted before completion is handled in the handleCompletedBuild as it will not be considered successful.